trace/sns/publisher.go

```diff
@@ -0,0 +1,67 @@
+// Package sns provides a set of common interfaces and structs for publishing messages to AWS SNS. Implementations
+// in this package also include distributed tracing capabilities by default.
+package sns
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/service/sns/snsiface"
+	"github.com/beatlabs/patron/errors"
+	"github.com/beatlabs/patron/trace"
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+)
+
+// Publisher is the interface defining an SNS publisher, used to publish messages to SNS.
+type Publisher interface {
+	Publish(ctx context.Context, msg Message) (messageID string, err error)
+}
+
+// TracedPublisher is an implementation of the Publisher interface with added distributed tracing capabilities.
+type TracedPublisher struct {
+	api snsiface.SNSAPI
+
+	// component is the name of the component used in tracing operations
+	component string
+	// tag is the base tag used during tracing operations
+	tag opentracing.Tag
+}
+
+// NewPublisher creates a new SNS publisher.
+func NewPublisher(api snsiface.SNSAPI) (*TracedPublisher, error) {
+	if api == nil {
+		return nil, errors.New("missing api")
+	}
+
+	return &TracedPublisher{
+		api:       api,
+		component: trace.SNSPublisherComponent,
+		tag:       ext.SpanKindProducer,
+	}, nil
+}
+
+// Publish tries to publish a new message to SNS. It also stores tracing information.
+func (p TracedPublisher) Publish(ctx context.Context, msg Message) (messageID string, err error) {
```

Comment:
There is a vital thing missing here: We do not propagate the tracing ID, which should be embedded in the attributes...

Reply:
I agree I totally missed this point. After looking at other implementations of the tracing, I added the missing information to the SNS publisher.

Here is what we see in Jaegger:

![image](https://user-images.githubusercontent.com/1004323/63276932-49c75e00-c29c-11e9-80c1-95557cdf7ad3.png)


It looks like a new trace is created for the SNS publisher only - I am not yet very familiar with tracing, is it an expected behaviour or am I missing something?

Reply:
Check out the kafka client https://github.com/beatlabs/patron/blob/master/trace/kafka/kafka.go.
The main thing here is to get the context, which contains the propagated tracing id from the calling code and create a new trace.
The second thing is to add this tracing id into the message headers (check createProducerMessage function)

Reply:
I managed to fix it and make it work - it requires to properly set up the bind between SNS and SQS (check example 4).

Regarding the trace title, unfortunately the topic ARN is very long as you can see. I removed the `publish` prefix but there's nothing much we can do about the name length.

Here is the final trace:
![image](https://user-images.githubusercontent.com/1004323/63363583-c70dd400-c36b-11e9-8948-e9c3ad138f06.png)